zerver/lib/zulip_update_announcements.py

```diff
@@ -0,0 +1,177 @@
+import os
+import re
+from datetime import timedelta
+from typing import List, Optional
+
+from django.conf import settings
+from django.db import transaction
+from django.db.models import Q, QuerySet
+from django.utils.timezone import now as timezone_now
+
+from zerver.actions.message_send import (
+    do_send_messages,
+    internal_prep_huddle_message,
+    internal_prep_stream_message,
+)
+from zerver.lib.message import SendMessageRequest
+from zerver.models.realm_audit_logs import RealmAuditLog
+from zerver.models.realms import Realm
+from zerver.models.users import UserProfile, get_system_bot
+
+zulip_update_announcements_file_path = os.path.join(
+    os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
+    "management/commands/zulip_update_announcements.md",
+)
+
+
+def get_latest_zulip_update_announcements_level() -> int:
+    with open(zulip_update_announcements_file_path) as zulip_updates:
+        content = zulip_updates.read()
+        match = re.search(r"\*\*Zulip update announcements level (\d+)\*\*", content)
+        assert match is not None
+        return int(match.group(1))
+
+
+def get_zulip_update_announcements_message_for_level(level: int) -> str:
+    with open(zulip_update_announcements_file_path) as zulip_updates:
+        content = zulip_updates.read()
+        current_level = f"**Zulip update announcements level {level}**"
+        previous_level = f"**Zulip update announcements level {level-1}**"
+        # Use regex to find the message between current_level and previous_level.
+        pattern = re.compile(
+            f"{re.escape(current_level)}(.*?)" + f"{re.escape(previous_level)}", re.DOTALL
+        )
+        match = re.search(pattern, content)
+        assert match is not None
+        return match.group(1).strip()
+
+
+def get_realms_behind_zulip_update_announcements_level(level: int) -> QuerySet[Realm]:
+    # Filter out deactivated realms. When a realm is later
+    # reactivated, send the notices it missed while it was deactivated.
+    realms = Realm.objects.filter(
+        Q(zulip_update_announcements_level__isnull=True)
+        | Q(zulip_update_announcements_level__lt=level),
+        deactivated=False,
+    ).exclude(string_id=settings.SYSTEM_BOT_REALM)
+    return realms
+
+
+def internal_prep_group_direct_message_for_old_realm(
+    realm: Realm, sender: UserProfile
+) -> Optional[SendMessageRequest]:
+    administrators = realm.get_human_admin_users()
+    emails = [administrator.email for administrator in administrators]
```

Comment:
Since, we already have `user_profiles`, so included `Addressee.for_user_profiles` to use.

Using `Addressee.for_user_ids` will lead to "fetch userprofiles/administrators"  -> "pass user IDs" -> "again fetch user profiles".